apptail/storage/tail_tracker.go

```diff
@@ -0,0 +1,173 @@
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Submit()
+	StartSubmissionTimer(retentionPeriod time.Duration)
+	IsInstanceRegistered(instKey string) bool
+	IsChildNodeInitialized(instKey string, childkey string) bool
+	GetFileCachedOffset(instkey string, fname string) int64
+}
+
+type BoxedInt64 struct{ V int64 }
+
+type TailNode map[string]*BoxedInt64
+
+type Tailer struct {
+	Instances map[string]TailNode
+}
+
+type tracker struct {
+	storage       Storage
+	Cached        *Tailer // do not expose this, it should ONLY be updated via Tracker methods
+	mux           *sync.Mutex
+	timerStopChan chan struct{} // used to send quit signal to timer
+}
+
+var (
+	MinIOTicker = 5 * time.Second
+)
+
+func NewTracker(s Storage) Tracker {
+	return &tracker{
+		storage: s,
+		mux:     &sync.Mutex{},
+		Cached: &Tailer{
+			Instances: make(map[string]TailNode),
+		},
+		timerStopChan: make(chan struct{}),
+	}
+}
+
+func (t *tracker) StartSubmissionTimer(retentionPeriod time.Duration) {
+	if retentionPeriod.Seconds() <= MinIOTicker.Seconds() {
+		seconds := retentionPeriod / (1000 * time.Millisecond)
+		log.Warnf("IMPORTANT: Setting retention period to %ds will increase your IO Rate", seconds)
+
+	}
+	ticker := time.NewTicker(retentionPeriod)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				t.Submit()
+			case <-t.timerStopChan:
+				ticker.Stop()
+				return
+			}
+
+		}
+	}()
+}
+
+func (t *tracker) RegisterInstance(instKey string) {
+	t.mux.Lock()
+	if _, instance_exist := t.Cached.Instances[instKey]; !instance_exist {
+		t.Cached.Instances[instKey] = TailNode{}
+		t.formatMap("Current Status")
+	}
+	t.mux.Unlock()
+}
+
+// this is mainly used for testing since we are not exposing Cached via interface
+func (t *tracker) IsInstanceRegistered(instKey string) bool {
+	var exist bool
+	t.mux.Lock()
+	if _, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		exist = instance_exist
+	}
+	t.mux.Unlock()
+	return exist
+}
+
+func (t *tracker) IsChildNodeInitialized(instKey string, childkey string) bool {
+	var exist bool
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childkey]; childNode_exist {
+			exist = childNode_exist
+		}
+	}
+	t.mux.Unlock()
+	return exist
+}
+
+func (t *tracker) InitializeChildNode(instKey string, childkey string, offSet int64) {
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childkey]; !childNode_exist {
+			tailNode[childkey] = &BoxedInt64{V: offSet}
+			t.Cached.Instances[instKey] = tailNode
+			t.formatMap("Current Status")
+		}
+	}
+	t.mux.Unlock()
+	runtime.Gosched()
+}
+
+func (t *tracker) GetFileCachedOffset(instkey string, fname string) int64 {
+	var offset int64
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instkey]; instance_exist {
+		offset = atomic.LoadInt64(&tailNode[fname].V)
```

Comment:
But this doesn't actually get us any benefit: things will mysteriously not work, since you don't handle the error condition.  I guess if you log it or something that's good enough? (The default panic behaviour includes logging, anyway...)